homeassistant/components/airvisual_pro/__init__.py

```diff
@@ -0,0 +1,136 @@
+"""The AirVisual Pro integration."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+from datetime import timedelta
+from typing import Any
+
+from pyairvisual import NodeSamba
+from pyairvisual.node import NodeConnectionError, NodeProError
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CONF_IP_ADDRESS,
+    CONF_PASSWORD,
+    EVENT_HOMEASSISTANT_STOP,
+    Platform,
+)
+from homeassistant.core import Event, HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.entity import DeviceInfo, EntityDescription
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
+
+from .const import DOMAIN, LOGGER
+
+PLATFORMS = [Platform.SENSOR]
+
+UPDATE_INTERVAL = timedelta(minutes=1)
+
+
+@dataclass
+class AirVisualProData:
+    """Define a data class."""
+
+    coordinator: DataUpdateCoordinator
+    node: NodeSamba
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up AirVisual Pro from a config entry."""
+    node = NodeSamba(entry.data[CONF_IP_ADDRESS], entry.data[CONF_PASSWORD])
+
+    try:
+        await node.async_connect()
+    except NodeProError as err:
+        raise ConfigEntryNotReady() from err
+
+    async def async_get_data() -> dict[str, Any]:
+        """Get data from the device."""
+        try:
+            data = await node.async_get_latest_measurements()
+        except NodeConnectionError as err:
+            hass.async_create_task(hass.config_entries.async_reload(entry.entry_id))
```

Comment:
Do we need to reload.  Can we reconnect instead?

Reply:
What if that reconnection fails? Seems like we get in a nested loop. Config entry reloading seemed appetizing because if upon entry reload, the connection fails, `ConfigEntryNotReady` is raised and a backoff-based retry loop is done for us.

I'll chew and see if any other ideas come forward.

Reply:
What you are doing here is likely fine. You could make it safer by checking if the reload task had already been created an only do it once to avoid multiple attempts to reload it in the rare event it doesn't unload before it fires the update again (users can request manual updates)

Reply:
https://github.com/home-assistant/core/pull/83869/commits/25452f693c83ee6734fba4103d56b40de0acf22c works well in my tests; let me know what you think.